### ISSUE TITLE
2021 Refactor Tracking Issue

### ISSUE THREAD

**Issue by federico-terzi** [espanso/espanso#594]
After 1.5 years since the start of this journey, it's clear how the current codebase is starting to show its limits. Espanso was my first Rust project and some of my early design choices (both caused by my lack of experience with the language and limited domain knowledge) were problematic to adapt to the new project requirements. So... It's finally time to refactor!

I'm planning to release a new major version (2.0.0), which means, *some* things might not be backward compatible for the sake of a cleaner/more robust implementation (although I'll try my best to maintain compatibility as much as possible). All development is happening in the `dev1.x` branch.

In this version, the project will heavily rely on cargo's workspaces to keep it modular and maintainable. This should reduce coupling and make it much easier to add/remove features in the future.

### Features
- [x] Implement new config handling logic to separate matches definition and app-specific configuration.
- [x] Implement the "legacy" config loader, to make espanso backward compatible with the old config format during the transition phase
- [x] Migration tool
- [x] Built-in patch system #310 
- [x] Search bar #420 
- [x] Experimental Regex triggers (with named groups to support arguments). 
- [X] Wayland support #287
- [X] Use native notifications on Windows
- [x] Module that handles the legacy/new directory mapping and migration. #611
- [x] Convert the macOS version into an App Bundle
- [x] Webview window to guide the Accessibility process on macOS
- [x] Implement new package system
- [x] Port espanso hub packages to new repository and format: https://github.com/espanso/hub

### Architectural changes
- [X] Extract the platform-specific application event-loop into a separate `ui` module
- [X] Extract the native input logic into a separate `detect` module
- [X] Extract the text injection logic into a separate `inject` module
- [x] Extract the clipboard logic into a separate `clipboard` module
- [x] Extract the application information extraction logic into a separate `app_info` module
- [x] Extract the macOS accessibility-related methods in a separate `accessibility` module
- [x] Extract the IPC logic into a separate `ipc` module
- [x] Extract the configuration parsing logic into a separate `config` module
- [X] Implement an alternative Linux subsystem based on the EVDEV and UINPUT, which is a necessary building block to port espanso to Wayland #287

### Deployment improvements
- [x] Create a `nightly` channel to rapidly ship bug fixes and improvements for testing, outside of the major release cycles.
- [x] Set up [clippy](https://github.com/rust-lang/rust-clippy) in the CI pipeline to fail when a warning is detected
- [x] Set up [rust fmt](https://github.com/rust-lang/rustfmt) in the CI pipeline to fail if the code is not properly formatted
- [x] Convert packager to use `cargo make`

### Other changes
- [X] Drop cmake requirement. The `cc` crate is more than capable of handling the C.C++/Obj-C compilation step and does not require CMake
- [x] Update the SECURITY.md file to reflect the changes in the architecture
- [x] Make sure all the bundled open-source projects ship with the necessary license

*Note: this roadmap will change over time*

This refactor will be the main focus of the project in the upcoming months, as having a solid core is a necessary building block for all the other features.
Once this one is ready and stable, we will move to other major features such as the GUI and improved package management.

If you have any ideas or concerns, feel free to write them below!

**Comment by 4lch4** [comment]
Hey @federico-terzi , gonna throw my hat in the ring rather shamefully and offer any assistance you can think of. I'm not yet capable of writing decent Rust code (still learning), but I've got plenty of code experience in other languages so I think I can do _something_ useful 😅

If anything, I'd be more than happy to help with testing across various environments, I enjoy automating that stuff quite a bit so just lemme know if I can help here or any other way. I'll look through issues and see if I can't find anything I feel I can tackle.

**Comment by 4lch4** [comment]
https://github.com/federico-terzi/espanso/issues/617#issuecomment-796702883

I see you mentioned needing to rewrite the docs. Give me a section you need to be rewritten and I'll get on it right away. I'd rather not pick something and work on it only to find out you've already done it.

**Comment by federico-terzi** [comment]
Hey @4lch4,
Thank you very much! 

> If anything, I'd be more than happy to help with testing across various environments, I enjoy automating that stuff quite a bit so just lemme know if I can help here or any other way. I'll look through issues and see if I can't find anything I feel I can tackle.

Would you like to help with the new CI pipeline? I'm planning to recreate it from scratch to account for the different requirements (nightly builds, new macOS format, ARM support, portable mode, etc). Let me know if you'd like to get involved in this process :)

> I see you mentioned needing to rewrite the docs. Give me a section you need to be rewritten and I'll get on it right away. I'd rather not pick something and work on it only to find out you've already done it.

Thank you, I appreciate that! I'm about halfway through the rewrite, and we should be able to start working on these new things (docs, pipeline, etc) in around a month or two. Is it ok for you if we start this conversation again then? :)

**Comment by 4lch4** [comment]
> Would you like to help with the new CI pipeline? I'm planning to recreate it from scratch to account for the different requirements (nightly builds, new macOS format, ARM support, portable mode, etc). Let me know if you'd like to get involved in this process :)

Would I!? Oh, Flying Spaghetti Monster, I'd absolutely love to. I do "unofficial" CICD support for the developers at my full-time job, so I'm rather familiar with the processes. Were you wanting to do it via GitHub Actions or stick to ADO (what I believe you're currently using)?

> Thank you, I appreciate that! I'm about halfway through the rewrite, and we should be able to start working on these new things (docs, pipeline, etc) in around a month or two. Is it ok for you if we start this conversation again then? :)

Hey, I'm here to help _you_ so whatever works best for you works for me :) We'll pick this up again whenever you're ready.

**Comment by federico-terzi** [comment]
> Would I!? Oh, Flying Spaghetti Monster, I'd absolutely love to. I do "unofficial" CICD support for the developers at my full-time job, so I'm rather familiar with the processes. Were you wanting to do it via GitHub Actions or stick to ADO (what I believe you're currently using)?

Awesome! I'd like to move all the pipeline to Github Actions (given that Github is now owned by Microsoft, I feel like Azure Pipelines won't have a long future...)

Sounds great! I'll let you know once we are ready to start experimenting :)

**Comment by srevinsaju** [comment]
@federico-terzi is it possible to test out the `dev-1.x` branch right now? I mean, manually building `espanso`. I tried building it and running the binary, but it seems like `espanso` changed a lot :sweat_smile: . Is it possible for me to participate in beta testing, as I am interested in testing out it in wayland :rocket: ✨

**Comment by federico-terzi** [comment]
@srevinsaju Thank you for the interest! Unfortunately, espanso can't be tested now because there is no `core` yet :)

The way the new version is structured is a series of modules for the various features (detection, injection, clipboard, etc) and then a "core" module that glues them all together... but there is no core yet :)

We currently have most of the modules implemented, but the core is now just a placeholder to test the various modules.
I'll make sure to let you all know when espanso 1.x is in an "alpha" state and thus ready to be tried

**Comment by srevinsaju** [comment]
Oh, cool! Thanks for your awesome work. Espanso is simply awesome  :fire: :sparkles:

**Comment by JeppeKlitgaard** [comment]
> I'm planning to release a new major version (2.0.0)

Should this be `1.0.0`, since the current stables versions are `0.7.x` - I might be missing something completely obvious!

Looking forward to seeing Espanso Core v1.0.0/v2.0.0!

**Comment by federico-terzi** [comment]
@JeppeKlitgaard Good point! It was 1.0.0 originally, but I recently decided to change it to v2.0 as it makes it clearer that there will be some breaking changes :)

(I'm working hard to make the transition as smooth as possible with the automatic migration tool and the compatibility mode)

Cheers :)

**Comment by JeppeKlitgaard** [comment]
@federico-terzi Makes perfect sense.

I think SemVer usually means that any minor releases before 1.0.0 can be breaking though and only after 1.0.0 is a change of major version needed to signal breaking changes. That being said, `espanso v2.0.0` does have a nice ring to it 😄.

**Comment by legendofmiracles** [comment]
I'd like to see a flag for turning off auto reloading, would it be possible to add this with the 2021 refactor? 
Thanks!

**Comment by federico-terzi** [comment]
Hey @legendofmiracles,

Of course! In fact, you should be able to do so already by adding `auto_restart: false` to your default.yml file

Cheers :)

**Comment by dansou901** [comment]
In the documentation at https://espanso.org/docs/next/install/linux/, there is the mention of an AppImage with download link. That AppImage doesn't exist yet, right?

**Comment by federico-terzi** [comment]
> In the documentation at https://espanso.org/docs/next/install/linux/, there is the mention of an AppImage with download link. That AppImage doesn't exist yet, right?

@dansou901 Yes, I'll try to publish the first binaries this weekend :) 

Last week I started writing the new docs. They are still far from being completed, but they should be enough for early testers to get started! The most important part is the migration guide: https://espanso.org/docs/next/migration/overview/

I'll let you know once Espanso v2 is ready to try!

Cheers :)

**Comment by dansou901** [comment]
Don't rush. Just wanted to know the status, I'll test it when it is ready.

**Comment by MattMattV** [comment]
Would you accept that I contribute to add `clippy` and `rust fmt` to CI and make them stop the current build in case of any warning ?

**Comment by federico-terzi** [comment]
@MattMattV Thank you for the help! There are a few warnings and formatting problems that might make that task difficult right now, so it's better to wait a couple of weeks until I fix the last ones after the release

**Comment by federico-terzi** [comment]
Hey guys,

I'm releasing the first official alpha v2 release later today (v2.0.1), which should be fairly usable already (I've been using it myself for a couple of weeks), but is likely to have some bugs and inconveniences :)

If anyone wants to give it a try (and enjoy the Searchbar and Regex trigger goodies 😎), here are the main takeaways:
* Documentation is still lacking, I just had the chance to write a [migration guide](https://espanso.org/docs/next/migration/overview/). That should be enough to get started with the new version, but there is a lot more to cover. Feel free to ask here on GitHub or Reddit if you don't understand something :) 
* Packages from the Hub are currently not working. With the great help of @4lch4, we are setting up the CI for the [new Hub](https://github.com/espanso/hub), which brings several improvements compared to the previous edition. In the meanwhile, you will need to manually copy the YAML files into your configs if you want to install a package. Sorry for the inconvenience :)
* Passive mode is currently not supported, though regex triggers might cover (and improve!) many of the possible use-cases
* Some commands, such as `espanso detect`, are yet to be ported. If you need to create an app-specific config, please see [this comment](https://github.com/federico-terzi/espanso/issues/762#issuecomment-936622335)
* The distribution format has changed for some platforms, please see the instructions for [macOS](https://espanso.org/docs/next/install/mac/), [Windows](https://espanso.org/docs/next/install/win/) and [Linux](https://espanso.org/docs/next/install/linux/). On macOS, Espanso is now a regular App bundle, which should make it much easier to deal with. On Linux, we still have a lot of work to do. For now, we have an AppImage available for X11 systems, with more formats and integrations (AUR, snap, etc) to come. Wayland users will need to compile Espanso from sources right now, I'm going to write a comment in the [wayland thread](https://github.com/federico-terzi/espanso/issues/287) with some instructions soon

Keep in mind that this is an alpha release, so things are likely to be buggy. Any feedback or bug report will be very helpful, so don't hesitate to let us know if you encounter any problems! :)

Happy expanding 🚀

**Comment by federico-terzi** [comment]
Hey guys!

I'm shipping the new 2.1.0-alpha release with support for [variable injection](https://espanso.org/docs/next/matches/variables/), which makes it possible to inject previous variables as parameters for other ones. For example, you can do things [like this](https://espanso.org/docs/next/matches/variables/#injecting-variables-inside-variables):

![image](https://user-images.githubusercontent.com/20475953/141647628-01052fe6-4647-41ed-aa84-5a40a2c00519.png)

That's really powerful and will open up many interesting use-cases. The only drawback is that it introduces a **breaking change** in the Form's control syntax, more info here: https://espanso.org/docs/next/migration/what_changed/#new-forms-control-syntax

If you are coming from version 0.7.3 to 2.1.0-alpha, the migration tool will take care of that for you. But if you are already using the v2, you'll need to convert all `{{control_name}}` to `[[control_name]]` in your form definitions. I'm sorry about that inconvenience, but I promise it's well worth it!

Cheers :)

**Comment by macintacos** [comment]
Hiya! I would like to preface my following question with the fact that I am _totally_ okay with waiting, and by no means am I asking for the development of this project to be rushed - I'm mostly just some guy who is very interested in using Espanso v2, but ideally doesn't want to use it while it is still under the "alpha" designation.

With all that being said my question: is there an ETA for the beta/stable release of Espanso v2? 🙂

**Comment by federico-terzi** [comment]
Hey @macintacos,

Thank you for the feedback! I definitely understand your point of waiting for a stable version before switching. From what I can tell, the current alpha version is pretty stable already, and the next release will likely be the first "beta" one :) So the wait shouldn't be too long

I would have loved to ship the first stable version last week, but I've been going through the busiest period of my life and I had to temporarily slow down the development to avoid getting burned out in the process :) That said, the project is far from being abandoned, I have huge plans for the future :) Just need some time to settle a few things in my life

Cheers :)

**Comment by AucaCoyan** [comment]
We merged #1836 and with that we finally did this! 🎉 
Closing as success!
